lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,859 @@
+    restarts: [],
+    dynamic: 0
+  ]
+
+  @doc false
+  defmacro __using__(opts) do
+    quote location: :keep do
+      @behaviour DynamicSupervisor
+      @opts unquote(opts)
+
+      @doc false
+      def child_spec(arg) do
+        default = %{
+          id: __MODULE__,
+          start: {__MODULE__, :start_link, [arg]},
+          type: :supervisor
+        }
+
+        Supervisor.child_spec(default, @opts)
+      end
+
+      defoverridable child_spec: 1
+
+      @doc false
+      def init(arg)
+    end
+  end
+
+  @doc """
+  Starts a supervisor with the given options.
+
+  The `:strategy` is a required option and the currently supported
+  value is `:one_for_one`. The remaining options can be found in the
+  `init/1` docs.
+
+  The `:name` option can also be used to register a supervisor name.
+  The supported values are described under the "Name registration"
+  section in the `GenServer` module docs.
+
+  If the supervisor is successfully spawned, this function returns
+  `{:ok, pid}`, where `pid` is the PID of the supervisor. If the supervisor
+  is given a name and a process with the specified name already exists,
+  the function returns `{:error, {:already_started, pid}}`, where `pid`
+  is the PID of that process.
+
+  Note that a supervisor started with this function is linked to the parent
+  process and exits not only on crashes but also if the parent process exits
+  with `:normal` reason.
+  """
+  @spec start_link(options) :: Supervisor.on_start()
+  def start_link(options) when is_list(options) do
+    keys = [:strategy, :max_seconds, :max_restarts, :max_children]
+    {sup_opts, start_opts} = Keyword.split(options, keys)
+    start_link(Supervisor.Default, init(sup_opts), start_opts)
+  end
+
+  @doc """
+  Starts a supervisor process with the given `module` and `arg`.
+
+  To start the supervisor, the `c:init/1` callback will be invoked in the given
+  `module`, with `arg` as its argument. The `c:init/1` callback must return a
+  supervisor specification which can be created with the help of the `init/1`
+  function.
+
+  If the `c:init/1` callback returns `:ignore`, this function returns
+  `:ignore` as well and the supervisor terminates with reason `:normal`.
+  If it fails or returns an incorrect value, this function returns
+  `{:error, term}` where `term` is a term with information about the
+  error, and the supervisor terminates with reason `term`.
+
+  The `:name` option can also be given in order to register a supervisor
+  name, the supported values are described in the "Name registration"
+  section in the `GenServer` module docs.
+  """
+  @spec start_link(module, term, GenServer.options()) :: Supervisor.on_start()
+  def start_link(mod, args, opts \\ []) do
+    GenServer.start_link(__MODULE__, {mod, args, opts[:name]}, opts)
+  end
+
+  @doc """
+  Dynamically adds a child specification to `supervisor` and starts that child.
+
+  `child_spec` should be a valid child specification. The child process will
```

Comment:
Looking at the signature, `child_spec` can't be provided in the new map format, which is also [officially called child specification](https://hexdocs.pm/elixir/Supervisor.html#module-child-specification).

Reply:
The map format is supported and we also have tests. `Supervisor.child_spec/2` is the function that supports it.